File: .omp/tools/_lib/gatelint.ts
// gate-lint (TS-порт tools/gate_lint.py) — детерминированный two-stage design-gate (INV-4).
// Модель-агностичен: код не рационализирует дефекты в PASS (в отличие от LLM-lead).
// Gate-1 (source↔spec): каждый SP несёт source-ref, резолвящийся в реальные непустые строки task-file.
// Gate-2 (spec↔ADR): каждый SP покрыт ≥1 AD.covers ИЛИ gap:true; каждый covers-id существует в spec.
// Gate-3 (EARS, P9 B.1a): FR-* (функц-требование) несёт source-ref + валидный EARS-шаблон (ядро-3
//   ubiquitous/event/unwanted). FR УСЛОВЕН — 0 FR-блоков = OK (мета/констрейнт-задача); линтим только
//   имеющиеся. gate_lint = СИНТАКС EARS (структура), семантика (testable/тот ли паттерн) — на critic.
import * as fs from "node:fs";
import * as path from "node:path";

const BLOCK_RE = /```(spec|adr|fr)\s*\n([\s\S]*?)```/g;
const SRC_RE = /^([^#]+)#L(\d+)(?:-L?(\d+))?$/;

export interface GateResult {
  findings: string[];
  nSp: number;
  nAd: number;
  nFr: number;
}

/** EARS-синтакс FR-блока (ядро-3). Возвращает finding-msg или "" (валиден). */
function earsCheck(fr: Record<string, string>): string {
  const pat = (fr.pattern || "").toLowerCase().trim();
  const st = (fr.statement || "").trim();
  if (!st) return "нет statement";
  const shall = /\bshall\b/i.exec(st);
  if (!shall) return `EARS: нет "SHALL" (${JSON.stringify(st)})`;
  const response = st.slice(shall.index + shall[0].length).trim();
  if (!response) return "EARS: пустой response после SHALL";
  const before = st.slice(0, shall.index);
  switch (pat) {
    case "ubiquitous":
      if (/^\s*(WHEN|WHILE|IF)\b/i.test(st)) return "EARS ubiquitous: не должно быть WHEN/WHILE/IF в начале";
      return "";
    case "event": {
      if (!/^\s*WHEN\b/i.test(st)) return "EARS event: должно начинаться с WHEN";
      const trig = before.replace(/^\s*WHEN\b/i, "").replace(/\bthe\s+system\s*$/i, "").trim();
      if (!trig) return "EARS event: пустой trigger между WHEN и SHALL";
      return "";
    }
    case "unwanted": {
      if (!/^\s*IF\b/i.test(st)) return "EARS unwanted: должно начинаться с IF";
      if (!/\bTHEN\b/i.test(before)) return "EARS unwanted: нет THEN перед SHALL";
      const cond = before.replace(/^\s*IF\b/i, "").replace(/\bTHEN\b[\s\S]*$/i, "").trim();
      if (!cond) return "EARS unwanted: пустое условие между IF и THEN";
      return "";
    }
    default:
      return `EARS: неизвестный pattern ${JSON.stringify(pat)} (ожидается ubiquitous|event|unwanted)`;
  }
}

/** Вытащить fenced-блоки заданного вида, распарсить `key: value` строки. */
function parseBlocks(mdPath: string, kind: "spec" | "adr" | "fr"): Record<string, string>[] {
  const text = fs.readFileSync(mdPath, "utf8");
  const out: Record<string, string>[] = [];
  for (const m of text.matchAll(BLOCK_RE)) {
    if (m[1] !== kind) continue;
    const entry: Record<string, string> = {};
    for (let line of m[2].split("\n")) {
      // spec-source содержит '#Lx' → режем только трейлинг-коммент ' # ...'; adr — весь #-хвост
      line = kind === "adr" ? line.split("#", 1)[0].replace(/\s+$/, "") : line.replace(/\s+#\s.*$/, "").replace(/\s+$/, "");
      const idx = line.indexOf(":");
      if (idx < 0) continue;
      entry[line.slice(0, idx).trim()] = line.slice(idx + 1).trim();
    }
    if (entry.id) out.push(entry);
  }
  return out;
}

/** Резолвится ли source-ref в реальные непустые строки task-file. */
function checkSourceRef(ref: string, repoRoot: string): { ok: boolean; msg: string } {
  const mm = SRC_RE.exec(ref.trim());
  if (!mm) return { ok: false, msg: `source не в формате <file>#Lx[-Ly]: ${JSON.stringify(ref)}` };
  const f = path.resolve(repoRoot, mm[1]);
  if (!fs.existsSync(f)) return { ok: false, msg: `source-file не найден: ${mm[1]}` };
  // splitlines-семантика (паритет с python gate_lint.py): роняем один трейлинг-newline перед split,
  // иначе "a\n" даёт лишнюю пустую строку → расходится счётчик длины с .splitlines().
  const lines = fs.readFileSync(f, "utf8").replace(/\r?\n$/, "").split(/\r?\n/);
  const start = parseInt(mm[2], 10);
  const end = mm[3] ? parseInt(mm[3], 10) : start;
  if (start < 1 || end > lines.length || start > end)
    return { ok: false, msg: `диапазон L${start}-L${end} вне файла (всего ${lines.length} строк)` };
  if (!lines.slice(start - 1, end).join("\n").trim())
    return { ok: false, msg: `строки L${start}-L${end} пусты` };
  return { ok: true, msg: "" };
}

/** Прогоняет two-stage gate. findings.length===0 → PASS. */
export function gateLint(taskPath: string, specPath: string, adrPath: string, repoRoot: string): GateResult {
  for (const p of [taskPath, specPath, adrPath])
    if (!fs.existsSync(p)) return { findings: [`файл не найден: ${p}`], nSp: 0, nAd: 0, nFr: 0 };

  const sps = parseBlocks(specPath, "spec");
  const ads = parseBlocks(adrPath, "adr");
  const frs = parseBlocks(specPath, "fr");   // FR живёт в spec (problem-space: WHAT система делает)
  const findings: string[] = [];
  if (!sps.length) findings.push("SPEC: не найдено ни одного ```spec-блока (формат нарушен)");
  if (!ads.length) findings.push("ADR: не найдено ни одного ```adr-блока (формат нарушен)");

  // Gate-1: SP ↔ source
  const spIds = new Set<string>();
  for (const sp of sps) {
    spIds.add(sp.id);
    const src = sp.source || "";
    if (!src) { findings.push(`G1 ${sp.id}: нет source-ref`); continue; }
    const r = checkSourceRef(src, repoRoot);
    if (!r.ok) findings.push(`G1 ${sp.id}: ${r.msg}`);
  }

  // Gate-2: SP ↔ AD coverage
  const covered = new Set<string>();
  for (const ad of ads) {
    const covers = (ad.covers || "").split(",").map((c) => c.trim()).filter(Boolean);
    if (!covers.length) findings.push(`G2 ${ad.id}: covers пуст (решение ничего не покрывает)`);
    for (const c of covers) {
      if (!spIds.has(c)) findings.push(`G2 ${ad.id}: covers ссылается на несуществующий ${c} (фантом)`);
      else covered.add(c);
    }
  }
  for (const sp of sps)
    if (!covered.has(sp.id) && (sp.gap || "").toLowerCase() !== "true")
      findings.push(`G2 ${sp.id}: не покрыт ни одним AD и не помечен gap:true`);

  // Gate-3: FR ↔ source + EARS-синтакс (FR условен — пусто = OK, линтим только имеющиеся)
  for (const fr of frs) {
    const src = fr.source || "";
    if (!src) findings.push(`G1 ${fr.id}: FR нет source-ref`);
    else {
      const r = checkSourceRef(src, repoRoot);
      if (!r.ok) findings.push(`G1 ${fr.id}: ${r.msg}`);
    }
    const e = earsCheck(fr);
    if (e) findings.push(`G3 ${fr.id}: ${e}`);
  }

  return { findings, nSp: sps.length, nAd: ads.length, nFr: frs.length };
}

File: tools/gate_lint.py
#!/usr/bin/env python3
"""
gate_lint.py — детерминированный two-stage gate (INV-4). Модель-агностичен.

Зачем: LLM-lead делает движения гейта и рационализирует дефекты в PASS
(P2-probe: Qwen нашёл сфабрикованный SP-point и всё равно штампанул). Код не рационализирует.

Gate-1 (source ↔ spec): каждый SP-point несёт source-ref, который резолвится в реальные
                         непустые строки task-file. Отсутствие/битый ref → BOUNCE.
Gate-2 (spec ↔ ADR):     каждый SP покрыт ≥1 AD.covers ИЛИ помечен gap: true.
                         Каждый id в AD.covers существует в spec (нет фантомных). Иначе BOUNCE.

Формат артефактов (строгий, парсится без PyYAML) — fenced-блоки:

  ```spec
  id: SP-bounded-memory
  source: docs/tasks/foo.md#L20-L22
  statement: cache must not grow unbounded
  gap: false            # опц.; true = осознанно без покрытия
  ```

  ```adr
  id: AD-eviction
  covers: SP-bounded-memory, SP-stale-rejection
  statement: TTL-based eviction
  ```

Usage:
  python tools/gate_lint.py --task <task.md> --spec <spec.md> --adr <adr.md>
Exit: 0 = PASS, 1 = BOUNCE (дефекты), 2 = ошибка ввода.
"""
from __future__ import annotations
import argparse
import re
import sys
from pathlib import Path

# Windows-консоль по умолчанию cp1251 → форсим UTF-8, иначе не-ASCII глифы падают.
try:
    sys.stdout.reconfigure(encoding="utf-8")  # type: ignore[union-attr]
except Exception:
    pass

BLOCK_RE = re.compile(r"```(spec|adr|fr)\s*\n(.*?)```", re.DOTALL)
SRC_RE = re.compile(r"^(?P<file>[^#]+)#L(?P<start>\d+)(?:-L?(?P<end>\d+))?$")
SHALL_RE = re.compile(r"\bshall\b", re.IGNORECASE)


def parse_blocks(md_path: Path, kind: str) -> list[dict]:
    """Вытащить fenced-блоки заданного вида, распарсить `key: value` строки."""
    text = md_path.read_text(encoding="utf-8")
    out: list[dict] = []
    for m in BLOCK_RE.finditer(text):
        if m.group(1) != kind:
            continue
        entry: dict[str, str] = {}
        for line in m.group(2).splitlines():
            line = line.split("#", 1)[0].rstrip() if kind == "adr" else line
            # для spec/fr не режем # (source содержит #Lx); режем только трейлинг-коммент после ' # '
            if kind in ("spec", "fr"):
                line = re.sub(r"\s+#\s.*$", "", line.rstrip())
            if ":" not in line:
                continue
            k, v = line.split(":", 1)
            entry[k.strip()] = v.strip()
        if entry.get("id"):
            out.append(entry)
    return out


def ears_check(fr: dict) -> str:
    """EARS-синтакс FR-блока (ядро-3). Возвращает finding-msg или '' (валиден)."""
    pat = fr.get("pattern", "").lower().strip()
    st = fr.get("statement", "").strip()
    if not st:
        return "нет statement"
    shall = SHALL_RE.search(st)
    if not shall:
        return f"EARS: нет \"SHALL\" ({st!r})"
    response = st[shall.end():].strip()
    if not response:
        return "EARS: пустой response после SHALL"
    before = st[:shall.start()]
    if pat == "ubiquitous":
        if re.match(r"\s*(WHEN|WHILE|IF)\b", st, re.IGNORECASE):
            return "EARS ubiquitous: не должно быть WHEN/WHILE/IF в начале"
        return ""
    if pat == "event":
        if not re.match(r"\s*WHEN\b", st, re.IGNORECASE):
            return "EARS event: должно начинаться с WHEN"
        trig = re.sub(r"\bthe\s+system\s*$", "", re.sub(r"^\s*WHEN\b", "", before, flags=re.IGNORECASE), flags=re.IGNORECASE).strip()
        if not trig:
            return "EARS event: пустой trigger между WHEN и SHALL"
        return ""
    if pat == "unwanted":
        if not re.match(r"\s*IF\b", st, re.IGNORECASE):
            return "EARS unwanted: должно начинаться с IF"
        if not re.search(r"\bTHEN\b", before, re.IGNORECASE):
            return "EARS unwanted: нет THEN перед SHALL"
        cond = re.sub(r"\bTHEN\b[\s\S]*$", "", re.sub(r"^\s*IF\b", "", before, flags=re.IGNORECASE), flags=re.IGNORECASE).strip()
        if not cond:
            return "EARS unwanted: пустое условие между IF и THEN"
        return ""
    return f"EARS: неизвестный pattern {pat!r} (ожидается ubiquitous|event|unwanted)"


def check_source_ref(ref: str, repo_root: Path) -> tuple[bool, str]:
    """Резолвится ли source-ref в реальные непустые строки."""
    mm = SRC_RE.match(ref.strip())
    if not mm:
        return False, f"source не в формате <file>#Lx[-Ly]: {ref!r}"
    f = (repo_root / mm.group("file")).resolve()
    if not f.exists():
        return False, f"source-file не найден: {mm.group('file')}"
    lines = f.read_text(encoding="utf-8").splitlines()
    start = int(mm.group("start"))
    end = int(mm.group("end") or start)
    if start < 1 or end > len(lines) or start > end:
        return False, f"диапазон L{start}-L{end} вне файла (всего {len(lines)} строк)"
    body = "\n".join(lines[start - 1:end]).strip()
    if not body:
        return False, f"строки L{start}-L{end} пусты"
    return True, body


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--task", required=True)
    ap.add_argument("--spec", required=True)
    ap.add_argument("--adr", required=True)
    ap.add_argument("--repo-root", default=".")
    args = ap.parse_args()

    repo_root = Path(args.repo_root).resolve()
    spec_p, adr_p = Path(args.spec), Path(args.adr)
    for p in (Path(args.task), spec_p, adr_p):
        if not p.exists():
            print(f"ERROR: файл не найден: {p}", file=sys.stderr)
            return 2

    sps = parse_blocks(spec_p, "spec")
    ads = parse_blocks(adr_p, "adr")
    frs = parse_blocks(spec_p, "fr")   # FR живёт в spec (problem-space: WHAT система делает)
    findings: list[str] = []

    if not sps:
        findings.append("SPEC: не найдено ни одного ```spec-блока (формат нарушен)")
    if not ads:
        findings.append("ADR: не найдено ни одного ```adr-блока (формат нарушен)")

    # Gate-1: SP ↔ source
    sp_ids: set[str] = set()
    for sp in sps:
        sid = sp["id"]
        sp_ids.add(sid)
        src = sp.get("source", "")
        if not src:
            findings.append(f"G1 {sid}: нет source-ref")
            continue
        ok, msg = check_source_ref(src, repo_root)
        if not ok:
            findings.append(f"G1 {sid}: {msg}")

    # Gate-2: SP ↔ AD coverage
    covered: set[str] = set()
    for ad in ads:
        aid = ad["id"]
        covers = [c.strip() for c in ad.get("covers", "").split(",") if c.strip()]
        if not covers:
            findings.append(f"G2 {aid}: covers пуст (решение ничего не покрывает)")
        for c in covers:
            if c not in sp_ids:
                findings.append(f"G2 {aid}: covers ссылается на несуществующий {c} (фантом)")
            else:
                covered.add(c)
    for sp in sps:
        sid = sp["id"]
        if sid not in covered and sp.get("gap", "").lower() != "true":
            findings.append(f"G2 {sid}: не покрыт ни одним AD и не помечен gap:true")

    # Gate-3: FR ↔ source + EARS-синтакс (FR условен — пусто = OK, линтим только имеющиеся)
    for fr in frs:
        fid = fr["id"]
        src = fr.get("source", "")
        if not src:
            findings.append(f"G1 {fid}: FR нет source-ref")
        else:
            ok, msg = check_source_ref(src, repo_root)
            if not ok:
                findings.append(f"G1 {fid}: {msg}")
        e = ears_check(fr)
        if e:
            findings.append(f"G3 {fid}: {e}")

    print(f"SP-points: {len(sps)} | AD-decisions: {len(ads)} | FR-points: {len(frs)} | findings: {len(findings)}")
    for f in findings:
        print("  [x] " + f)
    if findings:
        print("GATE: BOUNCE")
        return 1
    print("GATE: PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())
